pkg/streamingpromql/engine_fuzz_test.go

```diff
@@ -0,0 +1,238 @@
+func fixUpAndAssertPostingErrors(t *testing.T, prom, mimir *promql.Result) {
+	if mimir != nil && prom != nil && prom.Err != nil && prom.Err.Error() == "expanding series: unexpected all postings" {
+		assert.Equal(t, "unexpected all postings", mimir.Err.Error())
+		mimir.Err = prom.Err
+	}
+}
+
+// testInstantQueries executes the given query against both engines and asserts that the results match
+func testInstantQueries(t *testing.T, startT time.Time, query string, engines *queryEngines) {
+
+	prometheusQuery, prometheusError := engines.prometheus.NewInstantQuery(context.Background(), engines.queryable, nil, query, startT)
+	mimirQuery, mimirError := engines.mimir.NewInstantQuery(context.Background(), engines.queryable, nil, query, startT)
+
+	// if the Prometheus engine can not prepare the query then we expect the Mimir engine to also fail
+	if prometheusError != nil {
+		require.NotNil(t, mimirError)
+		return
+	}
+
+	require.NotNil(t, prometheusQuery)
+	require.NotNil(t, mimirQuery)
+
+	executeQueriesAndCompareResults(t, *engines, mimirQuery, prometheusQuery, query)
+}
+
+// testRangeQueries executes the given query against both engines and asserts that the results match
+func testRangeQueries(t *testing.T, startT time.Time, endT time.Time, step time.Duration, query string, engines *queryEngines) {
+	prometheusQuery, prometheusError := engines.prometheus.NewRangeQuery(context.Background(), engines.queryable, nil, query, startT, endT, step)
+	mimirQuery, mimirError := engines.mimir.NewRangeQuery(context.Background(), engines.queryable, nil, query, startT, endT, step)
+
+	// Note - prometheus validates time ranges and steps in http api handler, so it will not return an error at query preparation time
+	// Note - max resolution size (end-start/step) is validated in codec.go so we don't validate that here either
+	if prometheusError != nil || startT.After(endT) {
+		require.NotNil(t, mimirError)
+		return
+	}
+
+	require.NotNil(t, prometheusQuery)
+	require.NotNil(t, mimirQuery)
+
+	executeQueriesAndCompareResults(t, *engines, mimirQuery, prometheusQuery, query)
+}
+
+// doComparisonQuery invokes the given prepared queries against the given engines and validate that the results (inc series, errors, annotations) match.
+// engines also contains rules for how to compare the results (specifically rules about which queries we ignore annotation string mis-matches).
+func executeQueriesAndCompareResults(t *testing.T, engines queryEngines, mimirQuery promql.Query, prometheusQuery promql.Query, query string) {
+	// Execute our queries
+	defer prometheusQuery.Close()
+	prom := prometheusQuery.Exec(context.Background())
+
+	defer mimirQuery.Close()
+	mimir := mimirQuery.Exec(context.Background())
+
+	// Replace series Metric which are nil with an empty Labels{}
+	err := FixUpEmptyLabels(prom)
```

Comment:
Is this something that should move to `RequireEqualResults` perhaps? Why is this only a problem during fuzz testing?